### PR TITLE
[WIP][Feature][Flink] Upgrade flink to 1.14.6

### PR DESCRIPTION
**Pull request by liugddx**
## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)

### CODE REVIEW

**Comment by liugddx** [comment]
#3263

**Comment by liugddx** [comment]
#2291

**Comment by liugddx** [comment]
PTAL and give some suggestions, thanks @Hisoka-X @CalvinKirs @ashulin @EricJoy2048 @hailin0

File: seatunnel-apis/seatunnel-api-flink/src/main/java/org/apache/seatunnel/flink/FlinkEnvironment.java [hunk 1]
```diff
@@ -92,12 +81,8 @@ public CheckResult checkConfig() {
...
         // Batch/Streaming both use data stream api in SeaTunnel New API
-        createStreamEnvironment();
-        createStreamTableEnvironment();
-        if (!isStreaming()) {
-            createExecutionEnvironment();
-            createBatchTableEnvironment();
-        }
+        createEnvironment();
+        createTableEnvironment();
         if (config.hasPath("job.name")) {
```
Comment (wolfboys) [inline seatunnel-apis/seatunnel-api-flink/src/main/java/org/apache/seatunnel/flink/FlinkEnvironment.java:86]: I suggest that such generic parameters be extracted as constants

File: pom.xml [hunk 2]
```diff
@@ -129,7 +129,7 @@
         <logback.version>1.2.3</logback.version>
         <commons-logging.version>1.2</commons-logging.version>
         <iceberg.version>0.13.1</iceberg.version>
-        <flink.1.13.6.version>1.13.6</flink.1.13.6.version>
+        <flink.1.14.6.version>1.14.6</flink.1.14.6.version>
```
Comment (wolfboys) [inline pom.xml:132]: I don't know why `flink.1.14.6.version` should be defined so specifically,  Usually, the versions defined in maven's properties are simple and clear, e.g : `flink.114.version`

**Comment by TyrantLucifer** [comment]
This pull request involves upgrading and changing core functions, so it needs integration testing as comprehensive as possible, I will do integration tests in my local environment at free time, thank you for your contribution. It's an exciting feature for ST.

**Comment by liugddx** [comment]
Regarding the problem of smoothly upgrading flink, I want to specify the flink version by providing a parameter in env.like

```
env {
flink.version=1.13/1.14/1.15
}
```

In addition, different versions are divided by modules in the translation layer like spark.

image

**Comment by liugddx** [comment]
> Regarding the problem of smoothly upgrading flink, I want to specify the flink version by providing a parameter in env.like
> 
> ```
> env {
> flink.version=1.13/1.14/1.15
> }
> ```.In addition, different versions are divided by modules in the translation layer like spark.
> image
> ```

How about this?

@TyrantLucifer @hailin0 @Hisoka-X @CalvinKirs @ashulin @EricJoy2048

**Comment by liugddx** [comment]
> > Regarding the problem of smoothly upgrading flink, I want to specify the flink version by providing a parameter in env.like
> > ```
> > env {
> > flink.version=1.13/1.14/1.15
> > }
> > ```.In addition, different versions are divided by modules in the translation layer like spark.
> > image
> > ```
> 
> How about this?
> 
> @TyrantLucifer @hailin0 @Hisoka-X @CalvinKirs @ashulin @EricJoy2048

There are two solutions.

1. Specify the flink version when packaging(Limited flink version)

```
maven package -Dflink.version=1.13
```

2. Load the dependencies specified by SPI by getting the version number configured in env when running. This approach requires modification of the startup command.(The flink version can be specified dynamically. runtime)

**Comment by Hisoka-X** [comment]
Set `flink.version` in seatunnel-env.sh would be better.  Because this parameter alway same for single environment.

**Comment by liugddx** [comment]
> Set `flink.version` in seatunnel-env.sh would be better. Because this parameter alway same for single environment.

If users have multiple flink environments, can they be migrated to different versions at low cost?

**Comment by Hisoka-X** [comment]
> > Set `flink.version` in seatunnel-env.sh would be better. Because this parameter alway same for single environment.
> 
> If users have multiple flink environments, can they be migrated to different versions at low cost?

Of course, just change seatunnel-env.sh

**Comment by liugddx** [comment]
> > > Set `flink.version` in seatunnel-env.sh would be better. Because this parameter alway same for single environment.
> > 
> > 
> > If users have multiple flink environments, can they be migrated to different versions at low cost?
> 
> Of course, just change seatunnel-env.sh

I see.

**Comment by TyrantLucifer** [comment]
BTW, please revert codes about flink test e2e

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcSink.java [hunk 3]
```diff
@@ -131,7 +131,10 @@ public SeaTunnelDataType<SeaTunnelRow> getConsumedType() {
 
     @Override
     public Optional<Serializer<JdbcAggregatedCommitInfo>> getAggregatedCommitInfoSerializer() {
-        return Optional.of(new DefaultSerializer<>());
+        if (jdbcSinkOptions.isExactlyOnce()) {
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcSink.java:134]: Why change this? @ic4y PTAL

**Comment by Hisoka-X** [comment]
You upgrade flink to 1.14, but lost flink 1.13 support feature. Maybe you should extract some public methods, and then support different flink versions through different modules, the same as spark

**Comment by liugddx** [comment]
> You upgrade flink to 1.14, but lost flink 1.13 support feature. Maybe you should extract some public methods, and then support different flink versions through different modules, the same as spark

Yes, in has been discussed above, I will fit many versions of flink

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcSink.java (hunk 3 above)
Comment (liugddx) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcSink.java:134]: > Why change this? @ic4y PTAL
This code is to adapt to the higher version of flink.If you don't need two-phase commit, you need to ensure that the serializer is also empty.
https://github.com/apache/flink/blob/a921a4d848e70f1569c5f20efb7b634bdc80de82/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperatorFactory.java#L73

**Review by PraveenNanda124 — COMMENTED** [review]
Good work

**Comment by TyrantLucifer** [comment]
Because of #3963, this pull request will be closed.

**Pull request closed without merging**